src/main/java/net/minecraftforge/common/util/ITeleporter.java

```diff
@@ -19,49 +19,89 @@
      * by the computed cross-dimensional movement factor.
-     *
+     * <p>
      * Note that the supplied entity has not yet been spawned
      * in the destination world at the time.
      *
-     * @param entity           the entity to be placed
-     * @param currentWorld     the entity's origin
-     * @param destWorld        the entity's destination
-     * @param yaw              the suggested yaw value to apply
+     * @param entity the entity to be placed
+     * @param currentWorld the entity's origin
+     * @param destWorld the entity's destination
+     * @param yaw the suggested yaw value to apply
      * @param repositionEntity a function to reposition the entity, which returns the new entity in the new dimension. This is the vanilla implementation of the dimension travel logic. If the supplied boolean is true, it is attempted to spawn a new portal.
+     *
      * @return the entity in the new World. Vanilla creates for most {@link Entity}s a new instance and copy the data. But <b>you are not allowed</b> to create a new instance for {@link PlayerEntity}s! Move the player and update its state, see {@link ServerPlayerEntity#changeDimension(net.minecraft.world.server.ServerWorld, ITeleporter)}
      */
-    default Entity placeEntity(Entity entity, ServerWorld currentWorld, ServerWorld destWorld, float yaw, Function<Boolean, Entity> repositionEntity) {
-       return repositionEntity.apply(true);
+    default Entity placeEntity(Entity entity, ServerWorld currentWorld, ServerWorld destWorld, float yaw, Function<Boolean, Entity> repositionEntity)
+    {
+        return repositionEntity.apply(true);
     }
-    
-    //used internally to handle vanilla hardcoding
+
+    /**
+     * Is this teleporter the vanilla instance.
+     */
     default boolean isVanilla()
     {
-        return getClass() == Teleporter.class;
+        return this.getClass() == Teleporter.class;
+    }
+
+    /**
+     * Finds a portal to teleport to and creates a {@link TeleportationRepositioner.Result} for it. Defaults to calling vanilla Teleporter methods.
+     */
+    default Optional<TeleportationRepositioner.Result> findPortal(ServerWorld fromWorld, ServerWorld toWorld, Entity entity)
+    {
+    	Teleporter teleporter = toWorld.getDefaultTeleporter();
+        if (this instanceof Teleporter)
+            teleporter = (Teleporter) this;
+        return teleporter.func_242957_a(TeleporterHelper.getScaledPos(fromWorld, toWorld, new BlockPos(entity.getPositionVec())), toWorld.func_234923_W_() == World.field_234919_h_);
+    }
+
+    /**
+     * Creates a portal if one doesn't exist and returns the {@link TeleportationRepositioner.Result Result}. Defaults to calling vanilla Teleporter methods.
+     */
+    default Optional<TeleportationRepositioner.Result> createAndGetPortal(ServerWorld fromWorld, ServerWorld toWorld, Entity entity)
+    {
+    	Teleporter teleporter = toWorld.getDefaultTeleporter();
```

Comment:
Spacing is screwed up. Also is this changing the default of ITeleporter to spawning a portal instead of having a custom ITeleporter not have a portal? If so that is a breaking change from the logical standpoint even if not binary